File: src/karabo/tests/util/States_Test.cc
/* 
 * File:   States_Test.cc
 * Author: haufs
 *
 * Created on July 8, 2016, 10:08 AM
 */

#include "States_Test.hh"
#include <karabo/util/StateSignifier.hh>
#include <karabo/util/State.hh>

using namespace karabo::util;
using namespace std;

CPPUNIT_TEST_SUITE_REGISTRATION(States_Test);


States_Test::States_Test() {
}


States_Test::~States_Test() {
}


void States_Test::setUp() {

}


void States_Test::tearDown() {

}


void States_Test::testStringRoundTrip() {
    const State & s = State::CLOSED;
    const std::string & sstr = s.name();
    const State & s2 = State::fromString(sstr);
    CPPUNIT_ASSERT(s == s2);
}


void States_Test::testSignifier() {

    std::vector<State> s;
    s.push_back(State::DISABLED);
    s.push_back(State::COOLED);
    s.push_back(State::DECREASING);
    CPPUNIT_ASSERT(StateSignifier().returnMostSignificant(s) == State::DECREASING);
    s.push_back(State::RAMPING_UP);
    CPPUNIT_ASSERT(StateSignifier(State::ACTIVE, State::INCREASING).returnMostSignificant(s) == State::RAMPING_UP);
    CPPUNIT_ASSERT(StateSignifier().returnMostSignificant(s) == State::DECREASING);
    s.push_back(State::INTERLOCKED);
    CPPUNIT_ASSERT(StateSignifier().returnMostSignificant(s) == State::INTERLOCKED);
    s.push_back(State::UNKNOWN);
    CPPUNIT_ASSERT(StateSignifier().returnMostSignificant(s) == State::UNKNOWN);
}


void States_Test::testSignifierInitTrump() {

    std::vector<State> s;
    s.push_back(State::INIT);
    s.push_back(State::RUNNING);
    s.push_back(State::CHANGING);
    s.push_back(State::NORMAL);
    s.push_back(State::PASSIVE);
    s.push_back(State::ACTIVE);
    CPPUNIT_ASSERT(StateSignifier().returnMostSignificant(s) == State::INIT);
}


void States_Test::testInterlockTrump() {

    std::vector<State> s;
    s.push_back(State::RUNNING);
    s.push_back(State::CHANGING);
    s.push_back(State::NORMAL);
    s.push_back(State::PASSIVE);
    s.push_back(State::ACTIVE);
    s.push_back(State::INTERLOCKED);
    CPPUNIT_ASSERT(StateSignifier().returnMostSignificant(s) == State::INTERLOCKED);
}


void States_Test::testSignifierNonDefaultList() {

    std::vector<State> trumpList;
    trumpList.push_back(State::INTERLOCKED);
    trumpList.push_back(State::UNKNOWN);
    trumpList.push_back(State::KNOWN);

    std::vector<State> s;
    s.push_back(State::DISABLED);
    s.push_back(State::RUNNING);
    s.push_back(State::CHANGING);
    s.push_back(State::COOLED);
    s.push_back(State::DECREASING);
    s.push_back(State::UNKNOWN);
    s.push_back(State::INTERLOCKED);
    CPPUNIT_ASSERT(StateSignifier(trumpList).returnMostSignificant(s) == State::CHANGING);
}


void States_Test::testRunningTrumpActivePassive() {

    std::vector<State> s;
    s.push_back(State::DISABLED);
    s.push_back(State::RUNNING);
    s.push_back(State::ACTIVE);
    s.push_back(State::PASSIVE);
    CPPUNIT_ASSERT(StateSignifier().returnMostSignificant(s) == State::RUNNING);
}


void States_Test::testChainStates() {
    std::vector<State> s;
    s.push_back(State::ON);
    CPPUNIT_ASSERT(StateSignifier().returnMostSignificant(s) == State::ON);
    s.push_back(State::OFF);
    CPPUNIT_ASSERT(StateSignifier().returnMostSignificant(s) == State::OFF);
    s.push_back(State::ACQUIRING);
    CPPUNIT_ASSERT(StateSignifier().returnMostSignificant(s) == State::ACQUIRING);
    s.push_back(State::MOVING);
    CPPUNIT_ASSERT(StateSignifier().returnMostSignificant(s) == State::MOVING);
    s.push_back(State::INTERLOCKED);
    CPPUNIT_ASSERT(StateSignifier().returnMostSignificant(s) == State::INTERLOCKED);
    s.push_back(State::ERROR);
    CPPUNIT_ASSERT(StateSignifier().returnMostSignificant(s) == State::ERROR);
    s.push_back(State::INIT);
    CPPUNIT_ASSERT(StateSignifier().returnMostSignificant(s) == State::INIT);
    s.push_back(State::UNKNOWN);
    CPPUNIT_ASSERT(StateSignifier().returnMostSignificant(s) == State::UNKNOWN);
}


void States_Test::testComparisons() {
    CPPUNIT_ASSERT(State::CHANGING.isDerivedFrom(State::NORMAL)); // direct parentage
    CPPUNIT_ASSERT(!State::NORMAL.isDerivedFrom(State::CHANGING)); // direct parentage the other way round should not compare
    CPPUNIT_ASSERT(State::RUNNING.isDerivedFrom(State::NORMAL)); // direct parentage
    CPPUNIT_ASSERT(!State::CHANGING.isDerivedFrom(State::ERROR)); // no parentage
    CPPUNIT_ASSERT(!State::ERROR.isDerivedFrom(State::CHANGING)); // the other way round
    CPPUNIT_ASSERT(State::HEATED.isDerivedFrom(State::NORMAL)); // longer list of ancestors
    CPPUNIT_ASSERT(!State::KNOWN.isDerivedFrom(State::INCREASING)); // longer list of ancestors the other way round should not compare

    const State state(State::fromString("ON"));
    CPPUNIT_ASSERT(state == State::ON);
    CPPUNIT_ASSERT(state != State::INIT);

}

File: src/pythonKarabo/karabo/common/tests/states_test.py
import unittest

from karabo.common.states import State, StateSignifier


class States_TestCase(unittest.TestCase):
    def test_states_signifier(self):
        s = [State.DISABLED, State.COOLED, State.DECREASING]
        signifier = StateSignifier()
        self.assertIs(signifier.returnMostSignificant(s), State.DECREASING)
        s.append(State.RAMPING_UP)
        self.assertIs(signifier.returnMostSignificant(s), State.DECREASING)
        signifier = StateSignifier(staticMoreSignificant=State.ACTIVE,
                                   changingMoreSignificant=State.INCREASING)
        self.assertIs(signifier.returnMostSignificant(s), State.RAMPING_UP)
        s.append(State.INTERLOCKED)
        self.assertIs(signifier.returnMostSignificant(s), State.INTERLOCKED)
        s.append(State.UNKNOWN)
        self.assertIs(signifier.returnMostSignificant(s), State.UNKNOWN)

    def test_states_signifier_non_def_list(self):
        trumpList = [State.INTERLOCKED, State.UNKNOWN, State.KNOWN]
        s = [State.DISABLED, State.CHANGING, State.COOLED, State.DECREASING,
             State.UNKNOWN, State.INTERLOCKED]
        signifier = StateSignifier(trumplist=trumpList)
        self.assertIs(signifier.returnMostSignificant(s), State.CHANGING)

    def test_states_round_trip(self):
        s = State.ROTATING_CLK.name

        self.assertEqual(s, "ROTATING_CLK")
        s2 = State.fromString(s)
        self.assertIs(s2, State.ROTATING_CLK)

    def test_hierarchy(self):
        # direct parentage
        self.assertTrue(State.CHANGING.isDerivedFrom(State.NORMAL))
        # direct parentage the other way round
        self.assertFalse(State.NORMAL.isDerivedFrom(State.CHANGING))
        # no parentage
        self.assertFalse(State.CHANGING.isDerivedFrom(State.ERROR))
        # the other way round
        self.assertFalse(State.ERROR.isDerivedFrom(State.CHANGING))
        # longer list of ancestors
        self.assertTrue(State.HEATED.isDerivedFrom(State.NORMAL))
        # longer list of ancestors the other way round
        self.assertFalse(State.KNOWN.isDerivedFrom(State.INCREASING))

    def test_parent(self):
        for s in State:
            if s.parent is None:
                continue
            self.assertIsInstance(s.parent, State)
            self.assertEqual(s.value, s.name)

    def test_example(self):
        self.assertIs(State.OFF.parent, State.PASSIVE)
        self.assertNotEqual(State.OFF, State.OPENED)
        self.assertIs(State.INIT.parent, None)
        self.assertIs(State.STOPPED.parent, State.PASSIVE)

    def test_circlefree(self):
        for s in State:
            for i in range(10):
                s = s.parent
                if s is None:
                    break
            else:
                self.fail("all states must reduce to None")


if __name__ == '__main__':
    unittest.main()
